posts/2024-02-05-crates-io-status-codes.md

```diff
@@ -0,0 +1,35 @@
+---
+layout: post
+title: "crates.io: API status code changes"
+author: Tobias Bieniek
+team: the crates.io team <https://www.rust-lang.org/governance/teams/crates-io>
+---
+
+When cargo and crates.io were first implemented, [HTTP response status codes](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status) were ignored. cargo assumed that crates.io would always respond with a "[200 OK](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/200)" and any error messages are detected from the response JSON body. When crates.io responded with a different status code then cargo was showing the raw JSON body instead of a regular error message:
```

Comment:
> When cargo and crates.io were first implemented, [HTTP response status codes](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status) were ignored.

This is a bit uncharitable and overly blame-ey IMO-- there were absolutely places that the API *did* use semantically correct HTTP status codes, just not everywhere. I think it'd be nicer to say something like: 

> In the rush leading up to Rust 1.0's release, with the desire to also have a tool to manage dependencies and a registry where people could share code, there were places in the API between Cargo and crates.io that didn't return the correct [HTTP response status codes](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status). Then Rust's stability guarantees around backwards compatibility made this nontrivial to fix, as we wanted older versions of Cargo to continue working with the current crates.io API.

Reply:
> This is a bit uncharitable and overly blame-ey

I was struggling to find a way to express this in a blame-free way, but looks like I did not succeed 😅 

happy to rephrase it :)

Reply:
I don't have strong feelings on the exact wording, but the tense wanders a bit in the original paragraph, and the suggested version doesn't flow into the block after.

Reusing most of Carol's suggested wording but also still using this as an intro to the following output sample, how about:

> Cargo and crates.io were developed in the rush leading up to the Rust 1.0 release to fill the dual needs for a tool to manage dependencies and a registry that people could use to share code. This rapid work resulted in these tools being connected with an API that initially didn't return the correct [HTTP response status codes](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status). After the Rust 1.0 release, Rust's stability guarantees around backward compatibility made this non-trivial to fix, as we wanted older versions of Cargo to continue working with the current crates.io API.
>
> When an old version of Cargo receives a non-"[200 OK](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/200)" response, it displays the raw JSON body like this: